### PR TITLE
[Security] Deprecate the old authentication mechanisms

### PR DESCRIPTION
**Pull request by chalasr**
| Q             | A
| ------------- | ---
| Branch?       | 5.3
| Bug fix?      | no
| New feature?  | no
| Deprecations? | yes/
| Tickets       |  #39308 
| License       | MIT
| Doc PR        | todo

Now that the authenticator system proven working well and is considered stable, we can deprecate the old authentication listeners as well as the Guard component (+ integrations).

### CODE REVIEW

File: src/Symfony/Bundle/SecurityBundle/Resources/config/security_authenticator.php [hunk 1]
```diff
@@ -60,7 +59,9 @@
         ->alias(UserAuthenticatorInterface::class, 'security.user_authenticator')
 
         ->set('security.authentication.manager', NoopAuthenticationManager::class)
-        ->alias(AuthenticationManagerInterface::class, 'security.authentication.manager')
+        ->alias('AuthenticationManagerInterface::class', 'security.authentication.manager')
```
Comment (derrabus) [inline src/Symfony/Bundle/SecurityBundle/Resources/config/security_authenticator.php]: This change does not look right.

**Comment by chalasr** [comment]
Needs some work to make it green.

**Comment by nicolas-grekas** [comment]
Why 5.3? That's for 5.4 to me.

**Comment by chalasr** [comment]
Deprecating the old security system as a late change for 5.3 was discussed on slack with @wouterj and @fabpot.

**Comment by wouterj** [comment]
The reasoning is similar to #39308: I'm almost certain that we didn't yet test all uses of the old authentication system on the new one. So I want to buy us time to add the missing features/cases in 5.4, before removing the old system in 6.0. Deprecating the old system gives us the most chance to get this type of feedback from bigger applications.

**Comment by nicolas-grekas** [comment]
Got it, thanks.
(Rebase needed)

**Comment by chalasr** [comment]
This PR misses deprecating some service definitions, I'm going to finish it tonight.

**Comment by chalasr** [comment]
This is ready.

Note: this PR deprecates the main part of the old system, but it's not exhaustive: there are some classes which won't be needed in 6.0 that this PR does not deprecate. e.g. `NoopAuthenticationManager` is still needed internally for now. 
Why it's not an issue? Fixing the deprecations introduced here implies that you are not using the remaining to-be-deprecated stuff anymore, thus you won't be bothered for the same topic in the next minor.

**Review by fabpot — APPROVED** [review]

File: src/Symfony/Component/Security/Core/Authentication/AuthenticationManagerInterface.php [hunk 2]
```diff
@@ -13,12 +13,15 @@
 
 use Symfony\Component\Security\Core\Authentication\Token\TokenInterface;
 use Symfony\Component\Security\Core\Exception\AuthenticationException;
+use Symfony\Component\Security\Http\Firewall\X509AuthenticationListener;
```
Comment (wouterj) [inline src/Symfony/Component/Security/Core/Authentication/AuthenticationManagerInterface.php]: I don't think this addition is correct?

File: src/Symfony/Component/Security/Core/Authentication/AuthenticationManagerInterface.php [hunk 3]
```diff
@@ -13,12 +13,15 @@
...
 
 /**
  * AuthenticationManagerInterface is the interface for authentication managers,
  * which process Token authentication.
  *
  * @author Fabien Potencier <fabien@symfony.com>
+ *
+ * @internal since Symfony 5.3
```
Comment (wouterj) [inline src/Symfony/Component/Security/Core/Authentication/AuthenticationManagerInterface.php:23]: Is there something blocking us from deprecating this? Afaik, it's not used by the authenticator implementation

File: src/Symfony/Bundle/SecurityBundle/Resources/config/guard.php [hunk 4]
```diff
@@ -24,8 +24,10 @@
                 abstract_arg('stateless firewall keys'),
             ])
             ->call('setSessionAuthenticationStrategy', [service('security.authentication.session_strategy')])
+            ->deprecate('symfony/security-bundle', '5.3', 'The "%service_id%" service is deprecated, use the new authenticator system instead.')
 
         ->alias(GuardAuthenticatorHandler::class, 'security.authentication.guard_handler')
+        ->deprecate('symfony/security-bundle', '5.3', 'The "%alias_id%" alias is deprecated, use the new authenticator system instead.')
```
Comment (wouterj) [inline src/Symfony/Bundle/SecurityBundle/Resources/config/guard.php]:
```suggestion
            ->deprecate('symfony/security-bundle', '5.3', 'The "%alias_id%" alias is deprecated, use the new authenticator system instead.')
```

File: src/Symfony/Bundle/SecurityBundle/Resources/config/security_authenticator.php [hunk 5]
```diff
@@ -62,6 +62,7 @@
 
         ->set('security.authentication.manager', NoopAuthenticationManager::class)
         ->alias(AuthenticationManagerInterface::class, 'security.authentication.manager')
+        ->deprecate('symfony/security-bundle', '5.3', 'The "%alias_id%" alias is deprecated, use the new authenticator system instead.')
```
Comment (wouterj) [inline src/Symfony/Bundle/SecurityBundle/Resources/config/security_authenticator.php]:
```suggestion
            ->deprecate('symfony/security-bundle', '5.3', 'The "%alias_id%" alias is deprecated, use the new authenticator system instead.')
```

**Review by wouterj — COMMENTED** [review]
Wow, thanks Robin! I feel very sorry for you after seeing all the SecurityBundle tests that needed fixes as well..

>  Why it's not an issue? Fixing the deprecations introduced here implies that you are not using the remaining to-be-deprecated stuff anymore, thus you won't be bothered for the same topic in the next minor.

Is this also the reason for not having `trigger_deprecation()` calls in almost all Guard component classes?

File: src/Symfony/Component/Security/Core/Authentication/AuthenticationManagerInterface.php (hunk 3 above)
Comment (chalasr) [reply src/Symfony/Component/Security/Core/Authentication/AuthenticationManagerInterface.php:23]: It's implemented by the `NoopAuthenticationManager` so deprecating it now causes unfixable notices to end users

File: src/Symfony/Bundle/SecurityBundle/Resources/config/security_authenticator.php (hunk 5 above)
Comment (chalasr) [reply src/Symfony/Bundle/SecurityBundle/Resources/config/security_authenticator.php]: good catch, fixed! thanks

**Comment by chalasr** [comment]
> Is this also the reason for not having trigger_deprecation() calls in almost all Guard component classes?

@wouterj Nope, that was a mistake. It's fixed now.
The only files that have no `trigger_deprecation()` calls but only `@deprecated` annotations are abstract classes and interfaces as it causes such issues:

>.   1x: Since symfony/security-guard 5.3: The "Symfony\Component\Security\Guard\Authenticator\AbstractFormLoginAuthenticator" class is deprecated, use the new authenticator system instead.
    1x in ClassLoader::loadClass from Composer\Autoload
    
I asked @nicolas-grekas who confirmed that we have to skip triggering in this case.

Thanks for the review!

**Comment by fabpot** [comment]
Thank you @chalasr.

**Pull request merged**